Add Atom to Open With context menu list in Windows Explorer

raelyard opened the pull request:
This modifies the registry to include Atom in the context menu for the same file extensions for which it appears in the Open With context menu list in Finder on OS X.
#4893

raelyard committed:
Install Atom as option to open files

in the Windows Eplorer "Open with" context submenu

raelyard committed:
:checkered_flag: Add file extensions for open with

in Windows Explorer context menu

raelyard committed:
Fix lint problem with spaces following commas

Stanzilla commented:
:+1:

kevinsawicki commented on src/browser/squirrel-update.coffee:
```diff
@@ -21,6 +21,9 @@ fileKeyPath = 'HKCU\\Software\\Classes\\*\\shell\\Atom'
 environmentKeyPath = 'HKCU\\Environment'
+openWithKeyPath = 'HKCU\\Software\\Classes\\Atom'
+filetypeKeyPath = (fileExtension) -> "HKCU\\Software\\Classes\\.#{fileExtension}\\OpenWithProgIds"
+openWithFileExtensions = ['scpt', 'applescript', 'as', 'asp', 'asa', 'aspx', 'ascx', 'asmx', 'ashx', 'bib', 'c', 'cc', 'cp', 'cpp', 'cxx', 'c++', 'cs', 'coffee', 'COMMIT_EDITMSG', 'cfdg', 'clj', 'cljs', 'csv', 'tsv', 'cgi', 'fcgi', 'cfg', 'conf', 'config', 'htaccess', 'css', 'diff', 'dtd', 'dylan', 'erl', 'hrl', 'fscript', 'f', 'for', 'fpp', 'f77', 'f90', 'f95', 'h', 'pch', 'hh', 'hpp', 'hxx', 'h ...
```
This array seems pretty long, what do you think about storing this in a `.json` file and just doing:

```coffee
openWithFileExtensions = require './win32-open-file-extensions'
```

raelyard replied:
That's reasonable.  Thinking through the implications of making sure that file is included in the installation package and available in the path to be picked up in running specs.  Would be ideal to have one source that also works with the OS X plist, but might require more effort than it's worth.

raelyard committed:
:art: Move file extensions for installer to file

kevinsawicki commented on resources/win/win32-open-file-extensions.json:
```diff
@@ -0,0 +1,227 @@
+[
+"scpt",
```
What do you think about indenting these two spaces and also sorting this list alphabetically to make future additions easier?

kevinsawicki commented on src/browser/squirrel-update.coffee:
```diff
@@ -5,6 +5,7 @@ path = require 'path'
 appFolder = path.resolve(process.execPath, '..')
 rootAtomFolder = path.resolve(appFolder, '..')
 binFolder = path.join(rootAtomFolder, 'bin')
+binFolder = path.join(appFolder, 'resources', 'cli') if not fs.existsSync(binFolder)
```
We usually use `unless` instead of `if not`

kevinsawicki commented on src/browser/squirrel-update.coffee:
```diff
@@ -21,6 +22,9 @@ fileKeyPath = 'HKCU\\Software\\Classes\\*\\shell\\Atom'
 environmentKeyPath = 'HKCU\\Environment'
+openWithKeyPath = 'HKCU\\Software\\Classes\\Atom'
+filetypeKeyPath = (fileExtension) -> "HKCU\\Software\\Classes\\.#{fileExtension}\\OpenWithProgIds"
+openWithFileExtensions = require path.join(binFolder, 'win32-open-file-extensions.json')
```
I think it might be better to keep the `win32-open-file-extensions.json` in the `src/` folder somewhere instead of in `resources/` since it is a `.json` file and not a platform specific file type like a `.plist`.

raelyard replied:
Started moving this, but then thinking about it - it's not a platform specific file type, but it is a file for a specific platform.  There's also a .js file in the win folder.  .sh files as well (which are platform specific, but not Windows types).  Is the intent really to separate file types?  Will move it if desired, but wanted to ask the question.  If so, should it just go in the root of src or create a new subdirectory?

kevinsawicki replied:
> Will move it if desired

Yeah, I do think we should move it to somewhere in `src/`

> If so, should it just go in the root of src or create a new subdirectory?

I think that is a good place to put it :+1:

kevinsawicki commented on src/browser/squirrel-update.coffee:
```diff
@@ -72,9 +76,28 @@ installContextMenu = (callback) ->
         addToRegistry(args, callback)
 
+  recursivelyRegisterFileTypesForOpenWith = (index, callback) ->
+    if(index >= openWithFileExtensions.length)
```
:fire: No need for parens here

spongecattle replied:
:+1: Could help

kevinsawicki commented on src/browser/squirrel-update.coffee:
```diff
@@ -119,12 +142,27 @@ getPath = (callback) ->
+    spawnReg(['delete', keyPath, '/v', valueName, '/f'], callback)
+
+  recursivelyRemoveFileTypesForOpenWith = (index, callback) ->
+    if(index >= openWithFileExtensions.length)
```
:fire: no need for parens here

raelyard committed:
:art: Alphabatize and indent file extensions list

raelyard committed:
:fire: Remove parens for consistent syntax

raelyard committed:
:art: Use syntax consistent with conventions

raelyard committed:
Move Windows "open with" file extensions file

to src folder instead of resources because it's not a platform specific
file type

amytruong commented:
Did various testing around this. :+1:  @raelyard suggested downloading random selection of file types (especially media) to check the Open With options.

damieng commented:
I've merged the basic work on registering on https://github.com/atom/atom/pull/10818 so the application **can** be associated without any path/update issues.

However I think we should avoid actively associating with file extensions for now until we can provide opt-out options as people find t he current level of shell integration already disruptive https://github.com/atom/atom/issues/5901

lee-dohm commented:
Sounds good to me @damieng. Should this be closed then?

damieng commented:
Yes, let's close this for now and revisit specific pre-selected file types later.
